sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -2129,13 +2223,11 @@ std::unique_ptr<CurlNetworkConnection> CurlConnectionPool::ExtractOrCreateCurlCo
 // first connection to be picked next time some one ask for a connection to the pool (LIFO)
 void CurlConnectionPool::MoveConnectionBackToPool(
     std::unique_ptr<CurlNetworkConnection> connection,
-    HttpStatusCode lastStatusCode)
+    bool httpKeepAlive)
 {
-  auto code = static_cast<std::underlying_type<Http::HttpStatusCode>::type>(lastStatusCode);
-  if ((code < 200 || code >= 300) && lastStatusCode != HttpStatusCode::NotFound)
+  if (!httpKeepAlive)
   {
-    // A handler with previous response with Error can't be re-use.
-    return;
+    return; // The server has asked us to not re-use this connection.
```

Comment:
One more note: When the server returns 500 (Operation Timeout) or 503 (Server Busy) responses, does it set the "Connection: close" header? This is important because in such cases, the recommended application behavior is to reconnect to minimize the use of the (most likely) overloaded node.

Reply:
Thank you Mykhailo, this makes good sense. Let me to address this in a follow-up PR, so that the team members can review comment on that idea separately, while we don't hold this PR.

Reply:
> to reconnect to minimize the use of the (most likely) overloaded node.

Throttling doesn't necessarily mean the node is overloaded. It usually means the usage exceeds account limit. Theoretically all nodes share the same throttling configuration, so connecting to another node won't help.

Reply:
Sure about the throttling, but it doesn’t cover all cases of the 5xx errors. They may arise due to hardware or connectivity issues, occurring either on the storage or gateway nodes, or from unexpected load spikes generated by other tenants. Reconnecting can expedite issue resolution, as the balancer will likely select a healthier node.

Reply:
Do you think it makes sense to verify this on practice, and measure perf with 500/503 being auto-closed vs relying on Connection: close header, so that maybe it does not make a difference?

Reply:
FWIW, `Connection` header is in the allow-list for logging, so [if logging is on](https://github.com/Azure/azure-sdk-for-cpp/tree/main/sdk/core/azure-core#troubleshooting), it also could be possible to see what the values are.

Reply:
Well, it would be great to verify this situation in practice, but it's sooo rare that I doubt it can be properly reproduced. Throttling can be reproduced, but it's not the exact situation I'm worried about (the hardware issue with the node that wasn't caught on time). I would like to examine the behavior in other client libraries (for example, the .NET one). What do they do with the connection when a 5xx error is received?

Reply:
From what I can see, other language SDKs do not implement connection pools - they fully delegate it all to HTTP transport, kind of how we also do it here with WinHTTP transport. Any such transport is what you get with standard libs with the language, and it would not have any Azure-specific logic. I think it would implement standard handling of HTTP protocol, and would rely on `Connection` header in combination with HTTP version in the response, and would not take the HTTP response code into consideration.
If you want, you can try to find it yourself - https://github.com/orgs/Azure/repositories?language=&q=%22azure-sdk-for%22&sort=&type=public. I'll also ask people around me, on the occasion, and if I find anything interesting, I'll try to let you know.